pennylane/transforms/metric_tensor.py

```diff
@@ -400,27 +425,38 @@ def _metric_tensor_cov_matrix(tape, diag_approx):
     obs_list = []
     coeffs_list = []
     params_list = []
+    in_argnum_list = []
 
     for queue, curr_ops, param_idx, _ in graph.iterate_parametrized_layers():
         params_list.append(param_idx)
+        in_argnum_list.append([p in argnum for p in param_idx])
+
+        if not any(in_argnum_list[-1]):
+            # no tape needs to be created for this block
+            continue
+
         coeffs_list.append([])
         obs_list.append([])
 
         # for each operation in the layer, get the generator
-        for op in curr_ops:
-            obs, s = qml.generator(op)
-            obs_list[-1].append(obs)
-            coeffs_list[-1].append(s)
+        for p, op in zip(param_idx, curr_ops):
+            if p in argnum:
+                obs, s = qml.generator(op)
+                obs_list[-1].append(obs)
+                coeffs_list[-1].append(s)
 
         # Create a quantum tape with all operations
         # prior to the parametrized layer, and the rotations
         # to measure in the basis of the parametrized layer generators.
         with qml.queuing.AnnotatedQueue() as layer_q:
             for op in queue:
+                # TODO: Maybe there are gates that do not affect the
+                # generators of interest and thus need not be applied.

```

Comment:
I think the problem is precisely, that `graph.iterate_parametrized_layers` gives the `pre_ops` that would be necessary to apply all of the `ops` in the current layer, but since we might only be interested in a subset of the `ops` one can leave out certain elements `pre_ops` when they are not in the backwards light cone of the selection of operators from `ops`.

Reply:
Ohh, I see! 
That is a very nice catch! 
Spontaneously I'd think that it might be an option to change the parameters that are used to define "parametrized layers", so that the `graph` does this optimization for us. However, this is an additional optimization that doesn't need to be covered here, and can stay in as a "todo", in my opinion :)

On hardware, I'd hope that there is another compilation step that would notice that there are not generators measured after the `pre_ops`, skipping redundant gates.

Reply:
Ok, then I leave in the TODO.

Initially, I tried to solve the issue by constructing a new `graph` object with the subset of parametrized parameters specified by `argnum`. I didn't get this to work though, but I'm sure it's possible.